riscv-page-tables/src/page_table.rs

```diff
@@ -1296,3 +1333,65 @@ impl<'a, T: PagingMode> Drop for GuestStageMapper<'a, T> {
         }
     }
 }
+
+/// Defines a type of action an iterator should follow
+enum IteratorAction {
+    Proceed(PageSize, SupervisorPageAddr),
+    Skip(PageSize),
+    End,
+}
+
+/// Generate a supervisor page iterator for the given range
+struct SupervisorPageIter<
+    T: PagingMode,
+    F: FnMut(PageAddr<T::MappedAddressSpace>) -> IteratorAction,
+> {
+    current: PageAddr<T::MappedAddressSpace>,
+    end: PageAddr<T::MappedAddressSpace>,
+    update: F,
+}
+
+impl<T: PagingMode, F: FnMut(PageAddr<T::MappedAddressSpace>) -> IteratorAction>
+    SupervisorPageIter<T, F>
+{
+    pub fn new(
+        start: PageAddr<T::MappedAddressSpace>,
+        end: PageAddr<T::MappedAddressSpace>,
+        update: F,
+    ) -> Self {
+        Self {
+            current: start,
+            end,
+            update,
+        }
+    }
+}
+
+impl<T: PagingMode, F: FnMut(PageAddr<T::MappedAddressSpace>) -> IteratorAction> Iterator
+    for SupervisorPageIter<T, F>
+{
+    type Item = SupervisorPageAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.current >= self.end {
+                return None;
+            }
+
+            match (self.update)(self.current) {
+                IteratorAction::Proceed(page_size, paddr) => {
+                    self.current = self.current.checked_add_pages_with_size(1, page_size)?;
+                    if self.current > self.end {
+                        return None;
```

Comment:
is this case even possible to hit given the validation we did prior to constructing the iterator?

Reply:
No I agree it should never happen given we already ran all the needed checks beforehand.
But that means we assume this iterator must always be used after such validation has been performed.